sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/ContainerFeedRangesCache.scala

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.spark
+
+import com.azure.cosmos.{CosmosAsyncContainer, SparkBridgeInternal}
+import com.azure.cosmos.models.FeedRange
+import reactor.core.scala.publisher.SMono
+import reactor.core.scala.publisher.SMono.PimpJMono
+
+import java.time.Instant
+import java.time.temporal.ChronoUnit
+import scala.collection.concurrent.TrieMap
+// scalastyle:off underscore.import
+import scala.collection.JavaConverters._
+// scalastyle:on underscore.import
+
+private[spark] object ContainerFeedRangesCache {
+
+  private val cache = new TrieMap[String, CachedFeedRanges]
+
+  def getFeedRanges
+  (
+    container: CosmosAsyncContainer
+  ): SMono[List[FeedRange]] = {
+
+    val key = SparkBridgeInternal.getCacheKeyForContainer(container)
+
+    cache.get(key) match {
+      case Some(cached) =>
+        if (cached
+          .retrievedAt
+          .compareTo(Instant.now.minus(CosmosConstants.feedRangesCacheIntervalInMinutes, ChronoUnit.MINUTES)) >= 0) {
+
+          SMono.just(cached.feedRanges)
+        } else {
+          refreshFeedRanges(key, container)
+        }
+      case None => refreshFeedRanges(key, container)
+    }
+  }
+
+  private[this] def refreshFeedRanges(key: String, container: CosmosAsyncContainer): SMono[List[FeedRange]] = {
```

Comment:
Should this be `= synchronized {`? What if multiple threads land on it (get the existing cache but it's stale?

Reply:
Similar to https://github.com/Azure/azure-sdk-for-java/blob/main/sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosClientCache.scala#L90

Reply:
No - not needed here. CosmosClientCache needs to guard against creating two client instances when two threads would hit stale cache concurrently. In this case it doesn't matter whether multiple threads call getFeedRanges in the backend and update the cache simultanously. The underlying TrieMap is thread-safe and the additional I/O call to the backend is not harmful here.